File: services/lambda/domain-writer/handler.js
"use strict";

const { runPvdmWorkload } = require("./services/pvdm-runtime");
const { runIntegrityGate } = require("./lib/integrity-gate");

exports.handler = async (event) => {
  if (event.contract && !event.source_rows) {
    const gate = runIntegrityGate(event.contract);
    if (!gate.passed) {
      return { outcome: "verification_failed", message: "Integrity gate failed", gate };
    }
    return {
      outcome: "committed",
      workload_id: event.workload_id || "design-check",
      message: "Integrity gate passed; invoke with source_rows for PVDM",
      gate,
    };
  }

  const contract = event.contract || event.workload?.contract;
  if (!contract) {
    return { outcome: "verification_failed", message: "Missing contract in payload" };
  }

  const gate = runIntegrityGate(contract);
  if (!gate.passed) {
    return { outcome: "verification_failed", message: "Integrity gate failed", errors: gate.errors };
  }

  return runPvdmWorkload({
    contract,
    source_rows: event.source_rows || [],
    workload_id: event.workload_id || `wl-${Date.now()}`,
    resume_offset: event.resume_offset || 0,
  });
};

File: services/domain-writer/handler.py
"""
CogniMesh Python domain-writer entrypoint.

Honest outcomes only: never return a fake VRP PASS. Production PVDM runs in the
Node Lambda at services/lambda/domain-writer (runPvdmWorkload).
"""
from __future__ import annotations

import json
import os


def handler(event, context):
    contract = event.get("contract") or (event.get("workload") or {}).get("contract")
    if not contract:
        return {
            "outcome": "verification_failed",
            "vrp_verdict": "FAIL",
            "message": "Missing contract",
            "pattern": "vaquar-pvdm",
        }

    source_rows = event.get("source_rows") or (event.get("workload") or {}).get("source_rows") or []
    workload_id = event.get("workload_id") or (event.get("workload") or {}).get("workload_id") or "py-writer"
    resume_offset = int(event.get("resume_offset") or (event.get("workload") or {}).get("resume_offset") or 0)

    # Prefer Node PVDM unless an explicit full coordinator is available and invoked.
    try:
        from serverless_data_mesh import (  # noqa: F401
            DomainTransactionBoundary,
        )
    except ImportError:
        return {
            "outcome": "verification_failed",
            "workload_id": workload_id,
            "resume_offset": resume_offset,
            "vrp_verdict": "FAIL",
            "message": (
                "Python domain-writer does not simulate VRP PASS. "
                "Deploy services/lambda/domain-writer (Node PVDM) "
                "or install and wire a full serverless-data-mesh coordinator."
            ),
            "pattern": "vaquar-pvdm-stub",
            "row_count": len(source_rows),
        }

    # Package present but this handler still does not run IceGuard/VRP/catalog commit.
    return {
        "outcome": "verification_failed",
        "workload_id": workload_id,
        "resume_offset": resume_offset,
        "vrp_verdict": "FAIL",
        "message": (
            "serverless_data_mesh is importable but this handler does not execute "
            "Physical→Verify→Metadata. Use the Node domain-writer Lambda for production PVDM."
        ),
        "pattern": "vaquar-pvdm-unwired",
        "row_count": len(source_rows),
    }
